### PR TITLE
Tighten mypy 'strict optional' whitelist

### PR DESCRIPTION
**Pull request by danieleades**

### CODE REVIEW

File: sphinx/ext/autodoc/__init__.py [hunk 1]
```diff
@@ -1222,7 +1231,10 @@ class DocstringStripSignatureMixin(DocstringSignatureMixin):
     feature of stripping any function signature from the docstring.
     """
     def format_signature(self, **kwargs: Any) -> str:
-        if self.args is None and self.config.autodoc_docstring_signature:  # type: ignore
+        if (
+            self.args is None and
+            self.config.autodoc_docstring_signature  # type: ignore[attr-defined]
```
Comment (AA-Turner) [inline sphinx/ext/autodoc/__init__.py]:
```suggestion
            self.args is None
            and self.config.autodoc_docstring_signature  # type: ignore[attr-defined]
```

**Review by AA-Turner — APPROVED** [review]

File: sphinx/ext/autodoc/__init__.py [hunk 2]
```diff
@@ -1222,7 +1231,10 @@ class DocstringStripSignatureMixin(DocstringSignatureMixin):
     feature of stripping any function signature from the docstring.
     """
     def format_signature(self, **kwargs: Any) -> str:
-        if self.args is None and self.config.autodoc_docstring_signature:  # type: ignore
+        if (
+            self.args is None
+            and self.config.autodoc_docstring_signature  # type: ignore[attr-defined]
```
Comment (AA-Turner) [inline sphinx/ext/autodoc/__init__.py:1236]:
```suggestion
            and self.config.autodoc_docstring_signature  # type: ignore  # NoQA: W503
```

Comment (danieleades) [reply sphinx/ext/autodoc/__init__.py:1236]: is this still required? not seeing any lint errors after rebase

Comment (AA-Turner) [reply sphinx/ext/autodoc/__init__.py:1236]: The 'and' should start the line (per Knuth) -- I need to update the linting rules at some point.

A

Comment (danieleades) [reply sphinx/ext/autodoc/__init__.py:1236]: i can do it here if you like?

**Pull request merged**